src/wp-includes/global-styles-and-settings.php

```diff
@@ -255,3 +242,42 @@ function ( $item ) {
 		}
 	}
 }
+
+/**
+ * Private function to clean the caches under the theme_json group.
+ *
+ * @since 6.1.2
+ * @access private
+ */
+function _wp_clean_theme_json_caches() {
+	wp_cache_delete( 'wp_get_global_stylesheet', 'theme_json' );
+	WP_Theme_JSON_Resolver::clean_cached_data();
+}
+
+/**
+ * Tell the cache mechanisms not to persist theme.json data across requests.
+ * The data stored under this cache group:
+ *
+ * - wp_get_global_stylesheet
+ *
+ * There is some hooks consumers can use to modify parts
+ * of the theme.json logic.
+ * See https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/
+ *
+ * The rationale to make this cache group non persistent is to make sure derived data
+ * from theme.json is always fresh from the potential modifications done via hooks
+ * that can use dynamic data (modify the stylesheet depending on some option,
+ * or settings depending on user permissions, etc.).
+ *
+ * A different alternative considered was to invalidate the cache upon certain
+ * events such as options add/update/delete, user meta, etc.
+ * It was judged not enough, hence this approach.
+ * See https://github.com/WordPress/gutenberg/pull/45372
+ *
+ * @since 6.1.2
+ * @access private
+ */
+function _wp_add_non_persistent_theme_json_cache_group() {
+	wp_cache_add_non_persistent_groups( 'theme_json' );
```

Comment:
This should be here. See this example, https://github.com/WordPress/wordpress-develop/blob/acbbee8a1199f9ce70f1d3bbcc9cb80e32047e6c/src/wp-includes/load.php#L756. Add theme_json to this list.

Reply:
Also, for cross-reference and extra clarity. I'm not sure where this should be hooked. Using `plugins_loaded` was the way we came up with at https://github.com/WordPress/gutenberg/pull/46150#discussion_r1034624190 Is there a better place?

Reply:
Hmm, I wonder if it's better to move
```
add_action( 'plugins_loaded', '_wp_add_non_persistent_theme_json_cache_group' );
```
to `src/wp-includes/default-filters.php`.
System-wide hooks/actions are supposed to go into that file.
At the same time, `_wp_add_non_persistent_theme_json_cache_group` function can stay in `global-styles-and-settings.php`.

Reply:
Moved the hook to `default-filters.php` in https://github.com/WordPress/wordpress-develop/pull/3712/commits/ff7fd1e3932ad7b5ec8cd8cfa66ab6054e2de165 I had missed that, thanks for bringing it up!

Reply:
> I'm not sure where this should be hooked. Using plugins_loaded was the way we came up with at https://github.com/WordPress/gutenberg/pull/46150#discussion_r1034624190 Is there a better place?

Mostly depends on whether plugins may need it right away (while loading). 

The other question is: does it make sense for this to be removable by plugins? As far as I understand it, no.

As @spacedmonkey points out above the similar `wp_cache_add_non_persistent_groups( array( 'counts', 'plugins' ) );` is hard-coded and happens earlier. Thinking it makes sense for `wp_cache_add_non_persistent_groups( 'theme_json' );` to be there too. Then the helper function `_wp_add_non_persistent_theme_json_cache_group()` won't be needed and the code would be to just change the first one to: 
```
wp_cache_add_non_persistent_groups( array( 'counts', 'plugins', 'theme_json' ) );
```

Reply:
> The other question is: does it make sense for this to be removable by plugins? As far as I understand it, no.

That's a good point. :thinking: 

Is there a way for it to happen earlier but still be collocated in this file?

Alternatively, if it is moved elsewhere, how can we make sure people reading this code has clarity about it being non-persistent? I can only think of adding comments in every function that uses that group.

Reply:
> Is there a way for it to happen earlier but still be collocated in this file?

Don't think so unless there is a function that is called on including that file (which is generally a bad idea). Also not sure if that is particularly important. Thinking that the best option here is to add it at the bottom of `wp_start_object_cache()` which seems to be the best place for such settings, and add the explanation as inline comment. 

Many modern IDEs (probably all) would show cross references. Having a good explanation of what is being set and why, as in the current docblock, would be more than enough imho. It is easy to find. If you're worried that somebody may miss it, perhaps adding another inline comment (pointing to where it is set) the first time it is used in global-styles-and-settings.php would make it even easier.

Reply:
Done at https://github.com/WordPress/wordpress-develop/pull/3712/commits/45969f10d692f8cf83a6082e599dfcb790219118